crates/revm/revm-inspectors/src/tracing/builder/parity.rs

```diff
@@ -218,13 +235,134 @@ impl ParityTraceBuilder {
+                        let maybe_sub = match step.op.u8() {
+                            opcode::CALL
+                            | opcode::CALLCODE
+                            | opcode::DELEGATECALL
+                            | opcode::STATICCALL
+                            | opcode::CREATE
+                            | opcode::CREATE2 => sub_stack.pop_front().expect("missing sub trace"),
+                            _ => None,
+                        };
+
+                        instructions.push(Self::make_instruction(step, maybe_sub));
+                    }
+
+                    match current.parent {
+                        Some(parent) => {
+                            sub_stack.push_back(Some(VmTrace {
+                                code: Default::default(),
+                                ops: instructions,
+                            }));
+
+                            child_idx = child_idx_stack.pop().expect("missing child idx");
+
+                            current = self.nodes.get(parent).expect("missing parent");
+                        }
+                        None => break instructions,
+                    }
+                }
+            }
+        };
+
+        VmTrace { code: Default::default(), ops: instructions }
+    }
+
+    /// todo::n config
+    ///
+    /// Creates a VM instruction from a [CallTraceStep] and a [VmTrace] for the subcall if there is one
+    fn make_instruction(step: &CallTraceStep, maybe_sub: Option<VmTrace>) -> VmInstruction {
+        let maybe_storage = match step.storage_change {
+            Some(storage_change) => {
+                Some(StorageDelta { key: storage_change.key, val: storage_change.value })
+            }
+            None => None,
+        };
+
+        let maybe_memory = match step.memory.len() {
+            0 => None,
+            _ => {
+                Some(MemoryDelta { off: step.memory_size, data: step.memory.data().clone().into() })
+            }
+        };
+
+        let maybe_execution = Some(VmExecutedOperation {
+            used: step.gas_cost,
+            push: match step.new_stack {
+                Some(new_stack) => Some(new_stack.into()),
+                None => None,
+            },
+            mem: maybe_memory,
+            store: maybe_storage,
+        });
+
+        VmInstruction {
+            pc: step.pc,
+            cost: 0, // todo::n
+            ex: maybe_execution,
+            sub: maybe_sub,
+        }
+    }
 }
 
-/// Construct the vmtrace for the entire callgraph
-fn vm_trace(nodes: &[CallTraceNode]) -> VmTrace {
-    // TODO: populate vm trace
+/// addresses are presorted via DF walk thru [CallTraceNode]s
+///
+/// use recursion to fill the [VmTrace] code fields
+pub(crate) fn populate_vm_trace_bytecodes<DB, I>(
+    db: &DB,
+    depth_first_call_addresses: &mut I,
+    trace: &mut VmTrace,
+) -> Result<(), DB::Error>
+where
+    DB: DatabaseRef,
+    I: Iterator<Item = Address>,
```

Comment:
```suggestion
    I: IntoIterator<Item = Address>,
```

no need for the `.into_iter` call above, can just pass a vec directly